src/compiler/checker.ts

```diff
@@ -16867,31 +16846,33 @@ namespace ts {
                     }
                 }
                 else if (target.flags & TypeFlags.Index) {
-                    const targetType = (target as IndexType).type;
-                    // A keyof S is related to a keyof T if T is related to S.
-                    if (source.flags & TypeFlags.Index) {
-                        if (result = isRelatedTo(targetType, (<IndexType>source).type, /*reportErrors*/ false)) {
-                            return result;
+                    if (relation !== definitelyAssignableRelation) {

```

Comment:
Why does the definite assignment relation skip over relating `keyof S` and `keyof T`?

Reply:
If we have two `keyof` types, then `S` and `T` are distinct generic types. Since they're not the same type and generic types have no constraints under this relationship, seems there's no point in further exploration. Also, if I change to include this check we get circularity errors in the `circularlySimplifyingConditionalTypesNoCrash.ts` test plus error baseline changes in two other similar tests.

Reply:
What if we're relating `keyof S` to `keyof sub(S, S & T)`?

Reply:
Or even `keyof S` to `keyof Q[T]` where the constraint of `Q[T]` is the same as `S`?